### PR TITLE
Add context support for Hook TS types

### LINKED ISSUE #171: `context` property for interceptor is un-typed
The TS types for`.intercept()` doesn't include the `context` option, or the appropriate args in the subsequent `call`/`tap` handlers.

### PR DESCRIPTION
**Pull request by adierkens**
TL;DR

* Adds `context` support for the Hook interface 
* Ran `prettier` and adds `tapable.d.ts` to the list of files to format

--

Happy to split this PR up into 2 if that makes it easier to review.

The main change was the addition of another parameter to the base `Hook` interface for `Context`, that's then passed down to the `HookInterceptor` type, which is now a discriminated union based on the presence of the `context: true` parameter to the `intercept()` call and augments the expected arguments. 

I also updated the args in `call()` and `loop()` to leverage the argument types in their signature (instead of `any[]`) 

Fixes #171

### CODE REVIEW

File: tapable.d.ts [hunk 1]
```diff
@@ -86,12 +156,38 @@ declare class AsyncHook<T, R, AdditionalOptions = UnsetAdditionalOptions> extend
...
+export class AsyncParallelHook<
+	T,
+	AdditionalOptions = UnsetAdditionalOptions,
+	ContextType = unknown
+> extends AsyncHook<T, void, AdditionalOptions, ContextType> {}
+export class AsyncParallelBailHook<
+	T,
+	R,
+	AdditionalOptions = UnsetAdditionalOptions,
+	ContextType = unknown
```
Comment (alexander-akait) [inline tapable.d.ts:168]: Why it is unknown?

Comment (adierkens) [reply tapable.d.ts:168]: I wanted to avoid the use of `any` if possible. Context starts as an empty object, so I could use something like `Record<any, unknown>`
